src/iosrv/sched.c

```diff
@@ -26,73 +26,963 @@
+		.process_io = policy_fifo_process,
+	},
+	{	/* SCHED_POLICY_ID_RR */
+		.enqueue_io = NULL,
+		.process_io = NULL,
+	},
+	{	/* SCHED_POLICY_ID_PRIO */
+		.enqueue_io = NULL,
+		.process_io = NULL,
+	}
+};
+
+static void
+process_all(struct dss_xstream *dx)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	int			 rc;
+
+	if (info->si_req_cnt == 0) {
+		D_ASSERT(d_list_empty(&info->si_fifo_list));
+		return;
+	}
+
+	prune_purge_list(dx);
+	rc = d_hash_table_traverse(info->si_pool_hash, process_pool_cb, dx);
+	if (rc)
+		D_ERROR("XS(%d) traverse pool hash error. "DF_RC"\n",
+			dx->dx_xs_id, DP_RC(rc));
+
+	D_ASSERT(policy_ops[sched_policy].process_io != NULL);
+	policy_ops[sched_policy].process_io(dx);
+}
+
+static inline bool
+should_enqueue_req(struct dss_xstream *dx, struct sched_req_attr *attr)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+
+	if (sched_prio_disabled || info->si_stop)
+		return false;
+
+	D_ASSERT(attr->sra_type == SCHED_REQ_GC ||
+		 attr->sra_type == SCHED_REQ_IO ||
+		 attr->sra_type == SCHED_REQ_MIGRATE);
+
+	/* For VOS xstream only */
+	return dx->dx_main_xs;
+}
+
+static void
+req_enqueue(struct dss_xstream *dx, struct sched_request *req)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_req_attr	*attr = &req->sr_attr;
+	struct sched_req_info	*sri;
+
+	D_ASSERT(req->sr_pool_info != NULL);
+	sri = &req->sr_pool_info->spi_req_array[attr->sra_type];
+
+	D_ASSERT(d_list_empty(&req->sr_link));
+	if (attr->sra_type == SCHED_REQ_IO) {
+		D_ASSERT(policy_ops[sched_policy].enqueue_io != NULL);
+		policy_ops[sched_policy].enqueue_io(dx, req, NULL);
+	} else {
+		d_list_add_tail(&req->sr_link, &sri->sri_req_list);
+	}
+	req->sr_enqueue_ts = info->si_cur_ts;
+
+	sri->sri_req_cnt++;
+	info->si_req_cnt++;
+}
+
+int
+sched_req_enqueue(struct dss_xstream *dx, struct sched_req_attr *attr,
+		  void (*func)(void *), void *arg)
+{
+	struct sched_request	*req;
+
+	if (!should_enqueue_req(dx, attr))
+		return req_kickoff_internal(dx, attr, func, arg);
+
+	D_ASSERT(attr->sra_type < SCHED_REQ_MAX);
+	req = req_get(dx, attr, func, arg, ABT_THREAD_NULL);
+	if (req == NULL) {
+		D_ERROR("XS(%d): get req failed.\n", dx->dx_xs_id);
+		return -DER_NOMEM;
+	}
+	req_enqueue(dx, req);
+
+	return 0;
+}
+
+void
+sched_req_yield(struct sched_request *req)
+{
+	struct dss_xstream	*dx = dss_current_xstream();
+
+	D_ASSERT(req != NULL);
+	if (!should_enqueue_req(dx, &req->sr_attr)) {
+		ABT_thread_yield();
+		return;
+	}
+
+	D_ASSERT(req->sr_ult != ABT_THREAD_NULL);
+	req_enqueue(dx, req);
+
+	ABT_self_suspend();
+}
+
+void
+sched_req_sleep(struct sched_request *req, uint32_t msecs)
+{
+	struct dss_xstream	*dx = dss_current_xstream();
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_request	*tmp;
+	int			 rc;
+
+	D_ASSERT(req != NULL);
+	if (msecs == 0 || info->si_stop || req->sr_abort) {
+		sched_req_yield(req);
+		return;
+	}
+
+	/* Don't put GC & aggregation ULTs in sleep under space pressure */
```

Comment:
This check actually should be removed, it's up to caller to decide if the ULT needs be put into sleep. I'll fix it if the patch needs be refreshed or fix it in follow-on patches.